Sf request config

svenfuchs opened the pull request:
this is a follow up to https://github.com/travis-ci/travis-web/pull/2336, and it relies on:

* https://github.com/travis-ci/travis-web/pull/2393
* https://github.com/travis-ci/travis-web/pull/2397 and
* https://github.com/travis-ci/travis-web/pull/2398

so all commits from those 3 PRs are currently listed on this PR here, too. once those PRs are merged they won't any more, after then rebasing to master

svenfuchs commented:
the issue with repos that have (had) config validation turned off is the following:

production currently relies on checking the repo's `config_validation` setting in order to determine whether to display the validation result. however, in order to do that one has to be authenticated, even for public repos. this seems wrong. the validation result should be visible if the config is visible, so the setting is not a great way to do this. also, it is not accurate with regards to the history. one can run a build with the setting turned off, then turn the setting back on, and run another build. the UI would now show validation results for both builds.

the only historical artifact that is stored to the database are config validation messages. however, those will be just empty if either the setting was turned off (validation not run), or if the validation did not produce any messages because everything about the build config is entirely valid.

one other thing to consider is that we are going to force-enable build config validation for all repositories soon. how early exactly we can do that depends on getting this very feature here shipped: with the announcement blog post we will also announce force enabling build config validation after a grace period and some additional reminders).

it seems like an oversight that we haven't added a boolean flag like `requests.validated` to the database that would allow us to check if a build request was validated or not. however, even if we added this now we'd have about 3-4 months worth of requests where we could not accurately determine this after the fact any more.

so to me it seems like the best compromise to bite the bullet and simply always display the validation result, even if validation has not run on a historical build (or one of the few repos that still have this turned off intentionally).

piotr-travisci commented:
First of all, I would like to thank you for such a huge impact and great work 👏 
I have a few comments about UI.
You created automated format detection, which is absolutely great, but I feel the indicator of the config format looks a little bit like buttons. I think if we could add "Detected format" in front of the indicator it will cut off customers' questions.
<img width="606" alt="Screenshot 2020-03-27 at 16 36 51" src="https://user-images.githubusercontent.com/51313833/77773511-fdba7400-7049-11ea-94bd-ee92395f937f.png">
2. The search field in Branch selector does not work. Probably repo name is not passed properly, console output says 404 for URL https://api.travis-ci.com/repo/undefined/branches?name=sf&sort_by=name&limit=10&exists_on_github=true
It looks like repo.id (and repo) is undefined here https://github.com/travis-ci/travis-web/blob/d4f4feb0216d4a292d97484b399fd9125d3e7612/app/components/request/config-form.js#L40

svenfuchs commented on app/components/repo-show-tools.js:
```diff
@@ -2,15 +2,17 @@ import Component from '@ember/component';
 import { inject as service } from '@ember/service';
 import { computed } from '@ember/object';
-import { alias } from '@ember/object/computed';
+import { alias, reads } from '@ember/object/computed';
```
it was used on the `currentUser` property. i have changed it to use `reads` now: https://github.com/travis-ci/travis-web/commit/038501cdda0853ddd9850163c7c0cbde5a86740f

svenfuchs commented on app/components/request/configs.js:
```diff
@@ -0,0 +1,141 @@
+  classNames: ['request-configs'],
+  classNameBindings: ['status'],
+
+  preview: service('request-config'),
```
yes, i've renamed things back and forth a couple times. i have settled on this naming because i felt it best to name the service to be named consistently across all related components ... and the service itself and the component `request-config-preview` _also_ have a property `requestConfig` for the model that is returned from `/repo/:slug/request/config` ... so there's some naming conflicts here. 

i think once it is understood that `preview` is the service then `preview.rawConfigs`, `preview.requestConfig` etc reads quite well though, so i think this is an ok compromise.

svenfuchs commented on app/components/request/raw-config.js:
```diff
@@ -15,52 +16,46 @@ export default Component.extend({
     } catch (e) {
-      return config;
+      return this.config;
     }
```
nice, i've tried this out! however, the logic doesn't seem to be precisely the same, and it would incorrectly re-stringify a YAML string that JSON.parse has failed parsing: the current logic does not `JSON.stringify` the string if `JSON.parse` failed. it only does so if `JSON.parse` succeeded.

we can still get rid of the try/catch statement though:

```
    let object = parseWithDefault(this.config, null);
    return object ? JSON.stringify(object, null, 2) : this.config;
```

if the ternary is acceptable in this codebase then i think this is a huge improvement (i personally love ternaries ;) ...  without it we'd need an if statement, and i'm not sure if that's then more expressive or readable than just keeping the try/catch statement instead.

i have committed this for now: https://github.com/travis-ci/travis-web/commit/824283207d347de2ad65beea1c97e2392a811b9e

svenfuchs commented on app/mixins/components/can-trigger-build.js:
```diff
@@ -0,0 +1,18 @@
+  canCreateRequest: reads('repo.permissions.create_request'),
+  proVersion: reads('features.proVersion'),
+  enterpriseVersion: reads('features.enterpriseVersion'),
+  proOrEnterprise: or('proVersion', 'enterpriseVersion'),
```
hmmm, i'm not sure ... `version` for me would be a non-boolean value, wouldn't it? and it could have the value `org`, too, as opposed to `pro` (which should be `com`, but yeah) and `enterprise`.

imo the condition

```
this.canCreateRequest && (this.proOrEnterprise || !this.repoMigrated)
```

wouldn't read as well if we'd have

```
this.canCreateRequest && (this.version || !this.repoMigrated)
```

what do you think?

svenfuchs commented on app/mixins/components/with-config-validation.js:
```diff
@@ -11,9 +11,10 @@ export default Mixin.create({
+  // we want to show config messages for unauthenticated users, too. however,
+  // we then do not have a way to ask for the repository's config_validation
+  // setting. maybe we can check if the request.messages key is present?
+  showConfigValidation: true,
```
hmmmm, i don't think so. we would then not show the validation result on repos where one does not have a membership and there are no messages. e.g. if you look at the Ember repo's builds on Travis CI (assuming you're not a contributor to Ember) and if they have no validation messages, then the validation result would be hidden ... while it should message "everything valid, great job!"

there is also a longer explanation of the issue here: https://github.com/travis-ci/travis-web/pull/2399#issuecomment-604679907, in case you haven't seen it.

what do you think?

svenfuchs commented on app/helpers/incr.js:
```diff
@@ -0,0 +1,6 @@
+import { helper } from '@ember/component/helper';
```
i guess `incr` is just very understandable to me personally from other contexts :) 

i have renamed it: https://github.com/travis-ci/travis-web/commit/c4355434ee5c3d37a4594b4f648d5eb7f444b757

sorta commented on app/mixins/components/trigger-build.js:
```diff
@@ -0,0 +1,93 @@
+      yield timeout(searchDebounceRate);
+      yield this.showRequestStatus.perform(this.get('repo.id'), requestId);
+    }
+  }),
```
I notice that the branch searching task has the `restartable` task modifier, would we want that here as well? I think it may be necessary for the debouncing to work as expected.

And similarly, do we want any task modifiers on the other tasks in this file? Here's a good summary about them http://ember-concurrency.com/docs/task-concurrency

svenfuchs replied:
hmmm, i am not sure. i believe it should have the `drop` modifier ... once a the button has been pressed once, and the build request has been submitted, we want subsequent clicks to be ignored (which i believe also should happen by the way of disabling the button (which doesn't happen at the moment), ... but yeah, it might be good to also slap this modifier on the service here?

svenfuchs replied:
i've changed it to `drop` subsequent clicks ... and it seems that's doing the right thing. even if i click as hard and as quickly as i can i'll only ever manage to create a single request to that api endpoint: https://github.com/travis-ci/travis-web/pull/2399/commits/9fd4d69e87f394f0777c54231d13ed3a0c9f0221

sorta replied:
Sounds good, thanks!

svenfuchs commented on app/templates/components/request/config-form.hbs:
```diff
@@ -0,0 +1,134 @@
+        @changed={{action "change" "mergeMode"}}
+      >
+        Replace
+      </RadioButton>
```
interesting! i'll try that out. thank you!

svenfuchs replied:
works perfectly fine. cherrypicked onto this branch here. thank you!

sorta replied:
Excellent, thanks!

artursmirnov commented on app/adapters/request-config.js:
```diff
@@ -0,0 +1,25 @@
+import V3Adapter from 'travis/adapters/v3';
+
+export default V3Adapter.extend({
+  queryRecord(store, type, { id, data }) {
+    const url = `${this.buildURL('repo', id)}/request/config`;
+    return this.ajax(url, 'POST', { data });
+  },
```
This looks very much like [`createRecord`](https://api.emberjs.com/ember-data/3.12/classes/RESTAdapter/methods/createRecord?anchor=createRecord). It's doing [literally the same](https://github.com/emberjs/data/blob/v3.12.6/packages/adapter/addon/rest.js#L726-L732). Are there any specific reasons to use `queryRecord` here?

svenfuchs replied:
hmmm, are you sure? 

it seems the URL would be different, wouldn't it? `this.buildURL(type.modelName, null, snapshot, 'createRecord')` 

and `serializeIntoHash` seems to expect a model to be passed, not an arbitrary JS object https://github.com/emberjs/data/blob/v3.17.0/packages/serializer/addon/rest.js#L667

artursmirnov replied:
@svenfuchs you may be right about URL and serialization logic, but better approach (and "Ember way") would be to override appropriate API hooks and tweak as necessary. E.g. overriding [`buildURL`](https://api.emberjs.com/ember-data/3.12/classes/RESTAdapter/methods/buildURL?anchor=buildURL) hook would allow you to tweak URL.

svenfuchs replied:
@artursmirnov i've tried to make this work, but i'm just failing at it. i might be misunderstanding what you are saying about the appropriate API hooks, but my understanding is that `createRecord` expects to be passed an instance of a model, because `serializeIntoHash` does. we don't have such record instance though, and i wouldn't know what to pass here.

respectfully, i'd leave this as is and leave it for a more skilled person to refactor this later, if necessary at all.

artursmirnov replied:
@svenfuchs okay, let's leave it as is for now 👍 The reason why I pushed so hard on this is that current use of API is incorrect. `queryRecord` is aimed to **fetch** records from server, not to **save** data. This may not only be confusing for other devs and unexpected during debugging, but also may accidentally break after dependencies updates. We would have to refactor it later. I'll put it to my list  ✅ Thanks for trying! :)

artursmirnov commented on app/adapters/request-config.js:
```diff
@@ -0,0 +1,25 @@
+
+  handleResponse(status, headers, body) {
+    if (status == 400) {
+      let error = JSON.parse(body.error_message).error;
```
`JSON.parse` may throw error if it can't parse the string, and that would not be handled. Maybe worth wrapping to `try-catch` block?

svenfuchs replied:
we're controlling this on the api ourselves though ... so i wasn't sure it's needed? i can add it, of course.

svenfuchs replied:
done: https://github.com/travis-ci/travis-web/commit/44cd613529f3b30ff392d3bf9ce5330f35e9d0b2

artursmirnov replied:
@svenfuchs thank you! The only note that I have is that you can reuse existing helper: https://github.com/travis-ci/travis-web/blob/master/app/utils/json-parser.js

artursmirnov commented on app/components/job-tabs.js:
```diff
@@ -1,17 +1,36 @@
   isLog: not('isConfig'),
 
-  messages: reads('job.build.request.messages')
+  request: computed('job', 'build', function () {
+    if (this.build)
+      return this.build.get('request');
+    else
+      return this.job.get('build.request');
+  }),
```
It looks like the following should work too:

```suggestion
  request: or('build.request', 'job.build.request'),
```

svenfuchs replied:
woha, of course! 😆 

done: https://github.com/travis-ci/travis-web/commit/f20c4094c4311363723d2c94003c19f6ca636281

artursmirnov commented on app/components/request/config-form.js:
```diff
@@ -0,0 +1,69 @@
+  },
+
+  searchBranches: task(function* (query) {
+    const result = yield this.searchBranch.perform(this.get('repo.id'), query);
```
It shouldn't be necessary to use `get` here.

```suggestion
    const result = yield this.searchBranch.perform(this.repo.id), query);
```

svenfuchs replied:
i don't know why, but if i remove `get` here then the branch selector wouldn't make any api request at all, and never find any branches. if i keep it then it works ...

svenfuchs replied:
looking at this again i have found this ember warning in the log:

```
config-form.js:53 Error: Assertion Failed: You attempted to access the `id` property (of <(unknown):ember843>).
Since Ember 3.1, this is usually fine as you no longer need to use `.get()`
to access computed properties. However, in this case, the object in question
is a special kind of Ember object (a proxy). Therefore, it is still necessary
to use `.get('id')` in this case.
```

svenfuchs replied:
i suppose `request` is a proxy, and therefore `repo` is?

artursmirnov commented on app/components/request/config-form.js:
```diff
@@ -0,0 +1,69 @@
+  configMode: computed('config', function () {
+    const { config } = this;
+    const { JAVASCRIPT, YAML } = CONFIG;
+    return config && config.startsWith('{') ? JAVASCRIPT : YAML;
```
I'm not sure I'm getting it right, but string that starts with `{` looks more like `JSON` to me rather than `JAVASCRIPT`. Just wanted to bring your attention here, to make sure the logic is correct.

svenfuchs replied:
it seems the codemirror plugin requires the mode to be `javascript`, unfortunately, not `json` (there is no such file, and thus setting that mode fails ember's start up)

having said that, somewhere throughout all this refactoring branches and remerging/cherrypicking everything twice the setup for the editor modes was lost, and highlighting didn't work. i think i have fixed that now.

artursmirnov replied:
Ah, I see, it makes sense then, thank you for explaining 👍

artursmirnov commented on app/components/request/config-form.js:
```diff
@@ -0,0 +1,69 @@
+    let configs = this.get('request.uniqRawConfigs') || [];
+    configs = configs.reject(config => config.source === SOURCE.API);
+    if (this.config) {
+      configs.unshift({ config: this.config, source: SOURCE.API });
```
It's usually better to avoid using methods that mutate data, because they can accidentally leak unexpected changes to outer context. For example, if `configs.reject` a few lines above wouldn't create a copy of the array, this change would mutate `request.uniqRawConfigs` property, which, I guess, would not be expected.

Something like this would do the same, but without mutation:

```suggestion
      configs = [{ config: this.config, source: SOURCE.API }, ...configs];
```

svenfuchs replied:
done: https://github.com/travis-ci/travis-web/commit/4c287dff491425537ff83707eaeba37b4f0b977a

artursmirnov commented on app/components/request/configs-trigger-build-notice.js:
```diff
@@ -0,0 +1,23 @@
+import { pluralize } from 'ember-inflector';
+
+export default Component.extend({
+  tagName: 'p',
```
We're striving to make all new components tagless, because this concept has recently changed in Ember. In recent updates all Ember components are tagless, which means that `tagName` is always `''` (the property is deprecated and will be removed at some point), no bindings are made in component class, all markup and bindings are moved to template entirely (no wrapper element, template = DOM). The component class itself is not required anymore, by default Ember now generates only templates for components (so called "template-only components").

While old-style components still working properly, it makes sense to lean towards framework direction and use new recommended approaches whenever possible.

Some time soon we're going to update to the latest Ember version and would have to refactor all old-style components anyway.

svenfuchs replied:
interesting. i assume "tagless" does not mean "does not generate it's own tag", but just "does not produce tags other than `div`"?

done: https://github.com/travis-ci/travis-web/commit/4c3d455381b26d80640dee469b321936ac1e2b88

artursmirnov replied:
Right, "tagless" in Ember means that this component won't have a wrapper element. In order to make a component "tagless", it's necessary to set `tagName: ''` - to be empty string. Otherwise it'll default to `div` and will create a wrapper `<div>` element.

Here's a little bit more info:
![image](https://user-images.githubusercontent.com/4066489/78701236-5b36a680-790f-11ea-80e1-f31e8e790c89.png)
(posting as a screenshot because it doesn't allow to link to a specific section). Full post is here: https://blog.emberjs.com/2019/12/20/octane-is-here.html

artursmirnov replied:
In your fix it'll still create a wrapper `div`

svenfuchs replied:
thanks! 

done: https://github.com/travis-ci/travis-web/pull/2399/commits/97cd0a57cedebb20923b417440636d99809a866b

artursmirnov replied:
Nice! Thank you @svenfuchs!

svenfuchs commented on app/utils/computed.js:
```diff
@@ -0,0 +1,7 @@
+
+export default function truncate(property, chars) {
+  return computed(property, function () {
+    return this.get(property).slice(0, chars);
```
sounds good! i've made that:

```
    const value = this.get(property);
    return value ? value.slice(0, chars) : '';
```

https://github.com/travis-ci/travis-web/commit/f2303502e08da8dd8e898e365ae87e6fcf945d1d

artursmirnov commented on app/router.js:
```diff
@@ -61,6 +61,7 @@ Router.map(function () {
     this.route('build', { path: '/builds/:build_id', resetNamespace: true }, function () {
       this.route('config');
     });
+    this.route('config');
```
I think we have discussed the name of this route previously, but didn't come to a conclusion, right? Should we continue the discussion here?

svenfuchs replied:
yep, that's right. i've taken a note about this on this PR in the description above (the original discussion was [here](https://github.com/travis-ci/travis-web/pull/2405#discussion_r396682366). we can continue this discussion here, although i'm concerned others might not see it, but maybe that's alright, too.

the candidates mentioned so far are:

* `repo/:provider/:slug/config`
* `repo/:provider/:slug/request/config`
* `repo/:provider/:slug/build/config`
* `repo/:provider/:slug/requests/new/config`

i have given this more thought over the last two weeks, and i think we should stick with the current route name `repo/:provider/:slug/config` for these reasons:

* throughout our entire system "config" represents the _build config_ as provided by the user in their `.travis.yml` file, any config imports, and build config posted via API. for any other configuration we use the term "settings". this is the case in our database layout, internal model/service/etc naming, on our public API v2 and v3, and in reflected in the UI (as far as i can tell the only place "config" is being used visibly is on the build config tab.

  i think this addresses the concern of `repo/config` potentially being too overloaded in the future. i do not see us ever adding another concept of "repo config" which would not relate to the user's build config. in other words, i think in our system and product "config" is a synonymous with "build config".

* the user's build configuration that we accept represents an entire user interface on its own that has a significant complexity. this is, so far, not actually reflected in our UI (i suppose for the simple reason that it has been too complex to address, so far). with travis-yml in place we can now adjust that a little, and give build configs more of a first class visibility ... a proper "home" in our UI.
 
  i think this addresses the concern that `repo/config` is a rather top level, generic, not very much scoped naming. for such an important concern as the user's build config interface that seems alright to me. 

* the semantical mapping `repo <> config` directly matches the user's experience of storing a build config (their `.travis.yml`) within their git repository, which seems to make this a good choice, too.

* namings involving an action (such as "trigger build" or "new build request") seem to reduce the scope from an entity to an action, which makes them seem less fitting to me.

* naming involving `request` bring up the old dissent whether or not requests represent an internal implementation detail in our system. the argument made was that users care about builds, and can get confused by the term (build) request. while our API accepts build requests on the route `POST repo/:id/request` and i personally tend to disagree with the whole notion (i'd like requests to be promoted to be more first class citizens in the UI, too) ... i'd probably avoid that whole rabbit hole, and thus avoid the term `request` in a user facing route.

so, in conclusion ... i think `repo/:provider/:slug/config` is a great choice. the only other choice i think i would make would be `repo/:provider/:slug/build{s}/config`, but i don't really see a huge benefit added by mentioning `build`, too.

artursmirnov replied:
@svenfuchs thank you for such a detailed explanation! I totally agree with all of your points, and especially giving it another thought, I feel as well that `config = .travis.yml` and it belongs to `repo` directly, so it seems very natural and intuitive 👍

sorta replied:
The route and reasoning makes sense to me :+1:

artursmirnov commented on app/routes/repo/config.js:
```diff
@@ -0,0 +1,5 @@
+import TravisRoute from 'travis/routes/basic';
+
+export default TravisRoute.extend({
+  titleToken: 'Trigger Build',
```
The page title doesn't match the route name, which could be confusing. I'd suggest thinking about better title once we agree on the route name.

svenfuchs replied:
i agree, sounds good.

artursmirnov replied:
As long as the decision is made, would it make sense to make page title some thing like "Repository config"? What do you think @svenfuchs?

svenfuchs replied:
i've now made it `Repository build config`: https://github.com/travis-ci/travis-web/pull/2399/commits/a1c622350580d3e86212106a5a5f2710cc89b1b0

artursmirnov replied:
Neat 👍

svenfuchs commented on app/adapters/request-config.js:
```diff
@@ -0,0 +1,33 @@
+import AdapterError from '@ember-data/adapter/error';
+import V3Adapter from 'travis/adapters/v3';
+
+function parseJSON(string) {
+  try {
+    return JSON.parse(string);
+  } catch (e) {
+    return {};
+  }
+}
```
that one doesn't exist on this branch, yet. i'll have to rebase to master.

artursmirnov replied:
May be not rebase but just merge? Rebase here will likely be a nightmare 😅

svenfuchs replied:
just rebased. was pretty straight forward :)

done: https://github.com/travis-ci/travis-web/pull/2399/commits/3aeacafa9d09f100c8f9e64a2e28e703328e739e

artursmirnov replied:
Nice! Thank you 👍 Although, in your current use of the helper method it'll still break, because you didn't provide the default value to it: https://github.com/travis-ci/travis-web/commit/3aeacafa9d09f100c8f9e64a2e28e703328e739e#diff-4bf303175e8c929ff1750f67d6e9f125R13

It should be like this:
```js
let error = parseWithDefault(body.error_message, {}).error || {};
```
Or, with a little bit of javascript-fu 😄:
```js
let { error = {} } = parseWithDefault(body.error_message, {});
```
whatever you like more ;)

svenfuchs commented on app/components/request/configs-trigger-build-notice.js:
```diff
@@ -0,0 +1,22 @@
+  configs: 0,
+  status: null,
+  customizing: equal('status', 'customize'),
+  previewing: equal('status', 'preview'),
```
done: https://github.com/travis-ci/travis-web/pull/2399/commits/62b25716d854454001942578c72b064892176454

svenfuchs commented on app/components/request/configs.js:
```diff
@@ -0,0 +1,155 @@
+      },
+      type: 'api'
+    };
+  }),
```
you're right, great catch. done: https://github.com/travis-ci/travis-web/pull/2399/commits/eb801f9b1d386de01b8b44cc2c0893ca3bbd6116

svenfuchs commented on app/components/request/copy-button.js:
```diff
@@ -0,0 +1,30 @@
+  },
+
+  createClipboard() {
+    const clipboard = new window.ClipboardJS(this.element, { text: this.copyText });
```
makes sense. i think i just copied that from the copy button component that we're using, and didn't know i could import it instead.

done: https://github.com/travis-ci/travis-web/pull/2399/commits/56408f57875b1bbcc19d18deb2583ebe6d47e9f4

artursmirnov replied:
Nice, thank you! 😄

svenfuchs commented on app/components/request/copy-button.js:
```diff
@@ -0,0 +1,30 @@
+import Component from '@ember/component';
+import layout from '../../templates/components/request/copy-button';
```
done: https://github.com/travis-ci/travis-web/pull/2399/commits/7ef48be94508be6f8ab10165df6eab4d9d318a2a

svenfuchs commented on app/components/request/copy-button.js:
```diff
@@ -0,0 +1,30 @@
+import layout from '../../templates/components/request/copy-button';
+
+export default Component.extend({
+  layout: layout,
```
done: https://github.com/travis-ci/travis-web/pull/2399/commits/256b589b38d83ae75859bc6b9a9e49da0a37d03f

svenfuchs commented on app/components/request/message.js:
```diff
@@ -177,6 +177,46 @@ export default Component.extend({
+  },
+
+  invalid_ref(_, args) {
+    return args.message;
```
i understand that ideally the UI should be the party that constructs human readable messages.

the difference between these messages, and the messages above is that these are in response to API error conditions, while the messages above are validation messages that are produced throughout a successful validation process.

while i'd love to return structured data that the UI could then interpret on its own, API v3 does not provide a mechanism for returning such data on error conditions https://developer.travis-ci.com/resource/error#Error, and i think folding such change to API into this one here would extend the scope of this PR unproportionally.

since these are "just" error conditions i think what we have here is good enough. these messages are produced in travis-yml, and they should be rather readable. e.g. the error message format for this one here would be `invalid import reference: %s` with `%s` being the actual import reference provided by the user.

artursmirnov replied:
I see, @svenfuchs, thanks for the explanation! Let's keep it as is then 👍

artursmirnov commented on app/styles/app/modules/request/preview.scss:
```diff
@@ -0,0 +1,22 @@
+  .right {
+    flex-grow: 1;
+    align-items: stretch;
+    width: 50%;
```
For flex grid it should probably be `flex-basis`, shouldn't it?

svenfuchs replied:
hmm, i'm not sure. i've copied these from a generator site. i'll have a look.

svenfuchs replied:
turns out i could just remove the whole attribute: https://github.com/travis-ci/travis-web/pull/2399/commits/a9b2499df0d6c03693471fa1335407d28ca5f0de :)

artursmirnov replied:
@svenfuchs it now seems to me that you can remove the entire rule (`.left, .right`), as stretching child items to equal widths is the default behavior for flex. So, basically, if you just remove the entire block, the appearance should not change at all.

artursmirnov commented on app/styles/app/modules/request/buttons.scss:
```diff
@@ -0,0 +1,93 @@
+  vertical-align: bottom;
+  outline: none;
+  user-select: none;
+  -webkit-touch-callout: none;
+  -webkit-user-select: none;
```
You don't need to manually add vendor prefixes, Ember CLI's build pipeline includes Autoprefixer, which takes care of prefixes.

svenfuchs replied:
oh, nice, i didn't know! done: https://github.com/travis-ci/travis-web/pull/2399/commits/642d68d96413b78b6400f26d4f4322404a04f186

artursmirnov commented on app/styles/app/modules/request/form.scss:
```diff
@@ -0,0 +1,116 @@
+    margin-right: 6px;
+  }
+
+  .ember-radio-button {
+    margin-right: 16px;
+  }
```
As long as we don't use `ember-radio-button`, would it make sense to remove this style?

svenfuchs replied:
true, probably a left over from before the radio button PR. removed:

artursmirnov replied:
Thanks 👍

svenfuchs commented on app/components/request/configs.js:
```diff
@@ -0,0 +1,159 @@
+  branch: reads('originalBranch'),
+  message: reads('request.commit.message'),
+  config: reads('formattedApiConfig'),
+  mergeMode: reads('originalMergeMode'), // TODO store and serve merge mode for api request configs
```
ah, that's actually more of a note for myself. this won't be possible without some extra work in both Yml and Gatekeeper. 

i've removed it from here: https://github.com/travis-ci/travis-web/pull/2399/commits/642d68d96413b78b6400f26d4f4322404a04f186

artursmirnov replied:
That's okay to have TODO comments in the code on something that one has noticed but can't implement right now. So if you'd like to keep it, that's totally okay 👍 I just wanted to make sure, it's not aimed to be implemented in this PR :)

svenfuchs commented on tests/pages/trigger-build.js:
```diff
@@ -18,6 +18,12 @@ export default create({
+  configFormTriggerLinkIsHidden: isHidden('.option-dropdown .trigger-build-anchor'),
+  configFormTriggerLinkIsPresent: isPresent('.option-dropdown .trigger-build-anchor'),
+  showConfigForm: clickable('.option-dropdown .trigger-build-anchor'),
+  requestConfigButtonsIsPresent: isPresent('data-test-request-configs-button'),
```
done: https://github.com/travis-ci/travis-web/pull/2399/commits/642d68d96413b78b6400f26d4f4322404a04f186

svenfuchs replied:
done: https://github.com/travis-ci/travis-web/pull/2399/commits/642d68d96413b78b6400f26d4f4322404a04f186

svenfuchs commented on mirage/config.js:
```diff
@@ -659,6 +676,11 @@ export default function () {
+  this.get('commits/:id', (schema, { params }) => {
+    const commit = schema.commits.find(params.id);
+    return new Response(200, {}, commit);
+  });
```
done: https://github.com/travis-ci/travis-web/pull/2399/commits/642d68d96413b78b6400f26d4f4322404a04f186

artursmirnov replied:
Thanks!

artursmirnov commented on app/templates/components/request/config.hbs:
```diff
@@ -0,0 +1,34 @@
+        @language="yaml"
+        class="codedisplay line-numbers"
+        data-test-config="true"
+      >{{this.formattedConfig}}</CodeBlock>
```
I'd suggest to split it to multiple lines:
```suggestion
      >
        {{~this.formattedConfig~}}
      </CodeBlock>
```
The `~` should take care of whitespaces

svenfuchs replied:
i think that's what i had initially, but this would add extra whitespace around the formatted config, which would then show up in the code block (i.e. a `pre` tag), so i believe we need to keep it as is, unfortunately.

artursmirnov replied:
@svenfuchs are you sure you tried it with the `~`s around? They are crucial, as this is what tells Handlebars to remove whitespaces around the expression.

svenfuchs replied:
aaah, no, i haven't. i've missed those in your suggestion, too. 

done: https://github.com/travis-ci/travis-web/pull/2399/commits/56232b70163954d0ad566b23789dd2e7bd2b835e

artursmirnov commented on app/templates/components/request/configs.hbs:
```diff
@@ -0,0 +1,101 @@
+      />
+    {{/if}}
+
+    {{#if (or this.closed this.open)}}
```
This looks a bit confusing to me 😕 From how it reads, the condition should always be `true`. Just wanted to bring your attention, to make sure, the logic is correct here.

svenfuchs replied:
ah, i see what you mean, haha :)

hmmm, the component can be in the states `closed`, `open`, `customizing`, and `previewing`. i've gone back and forth between various combinations of namings, but couldn't come up with a better version than this.

we could of course move this to the component and give it a property `super-hard-to-name`, but i'm not sure that would be an improvement ...

i'd probably keep it as is?

artursmirnov replied:
Yea, it works perfectly, I just wanted to make sure it's not a mistake :)

artursmirnov commented on app/templates/components/request/configs.hbs:
```diff
@@ -0,0 +1,101 @@
+      />
+    {{/if}}
+
+    {{#if (or this.closed this.open)}}
```
Same here

svenfuchs replied:
see above

svenfuchs commented on app/mixins/components/trigger-build.js:
```diff
@@ -0,0 +1,93 @@
+          merge_mode: this.mergeMode
+        }
+      };
+      return yield this.api.post(`/repo/${this.get('repo.id')}/requests`, { data: data });
```
haha, ok, done: https://github.com/travis-ci/travis-web/pull/2399/commits/14f8b2d2edcb5ec7c31fc2ce469774dba75002be

artursmirnov replied:
👍

svenfuchs commented on app/services/request-config.js:
```diff
@@ -0,0 +1,42 @@
+    }).then((result) => {
+      this.set('loaded', true);
+      this.set('result', result);
+    });
```
ok, i have now made this: https://github.com/travis-ci/travis-web/pull/2399/commits/f2002bba375947ebe347d7e92d1c25eed441cc4b

artursmirnov replied:
Nice! 😄

artursmirnov commented:
Some visual feedback:

1) On the build page with a matrix view the config validation shows up, it's clickable, but doesn't show any content when clicking to expand:

![2020-04-09 at 22 24](https://user-images.githubusercontent.com/4066489/78933176-8b1daf80-7ab1-11ea-970d-74f3884ba28e.gif)


2) Please notice on the screencast above that `?` icon is not clickable.


3) Clicking any job on matrix view makes build header change, but nothing renders below it until page reload:

![2020-04-09 at 22 31](https://user-images.githubusercontent.com/4066489/78933538-257df300-7ab2-11ea-9b28-191f04daec52.gif)


4) Copy button doesn't have hover effect and thus doesn't look clickable. It looks a little bit weird:

![2020-04-09 at 22 36](https://user-images.githubusercontent.com/4066489/78933975-cb316200-7ab2-11ea-9f60-c3b2209f067e.gif)


5) On narrow screens (mobile) it doesn't look good:

![image](https://user-images.githubusercontent.com/4066489/78934455-9eca1580-7ab3-11ea-811e-61518369d434.png)


6) On [travis-web](https://sf-request-config.test-deployments.travis-ci.com/github/travis-ci/travis-web/builds/159678961/config) it doesn't seem to work stable, please see the screencast:

![2020-04-09 at 22 49](https://user-images.githubusercontent.com/4066489/78935016-a1793a80-7ab4-11ea-825d-9fa2ebae2589.gif)


7) After clicking "Trigger build" and the "Preview", I see empty configs on [travis-api](https://sf-request-config.test-deployments.travis-ci.com/github/travis-ci/travis-api/builds/159632571/config): 

![image](https://user-images.githubusercontent.com/4066489/78935624-c1f5c480-7ab5-11ea-8d42-0a0d6cdfcde5.png)


8) Refreshing the page on "Preview" (right after (7) above) makes me stuck on that view, being unable to disable or cancel the preview mode:

![2020-04-09 at 22 57](https://user-images.githubusercontent.com/4066489/78935731-fcf7f800-7ab5-11ea-82cf-476620ff0877.gif)

The only way to exit is to edit URL manually.

9) Erasing build commit in request form makes the request fail even though the form says that commit is optional:

![2020-04-09 at 23 06](https://user-images.githubusercontent.com/4066489/78936238-dbe3d700-7ab6-11ea-9067-6188f4451a99.gif)

With commit sha in place the same build config runs properly.

10) "Trigger build" button looks like a toggle to me. It shouldn't do different actions depending on the state (e.g. enable edit mode or trigger build). Usually, when a user clicks something, clicking on the same button either does the same or "unclicks" - does the opposite action. This is what users usually expect. Running a build by this button looks unexpected. I expect there will be a lot of builds started by accident caused by this behavior.

11) The behavior after triggering a build also looks a bit unexpected to me (redirecting to requests page). I'd suggest keeping user on the same page, show a flash message and give options what they can do next, e.g. see requests or open the build. Or just stay on the same page. Automatic redirect is distracting. What happened with the changes that I just made? Are they gone? Panic!

svenfuchs commented:
@artursmirnov thanks for the observations and notes! i think we'll have to give this another thorough click-through once we're done with all the reviews and such, but this is already very helpful.

> 1. empty messages list clickable but nothing happens

i have added another css class, and made it so that there's no cursor when there are no messages in the list (i had also removed the extra padding when the element was clicked/expanded). i think this is alright. adding a text "your config is all valid 🎉 " also would be possible, but i like what i now have.

> 2. Please notice on the screencast above that ? icon is not clickable.

errr, weird. any idea why that is and how to fix it? there's clearly a link, and it has a `href`. does the header steal and cancel the event?

> 3. Clicking any job on matrix view makes build header change, but nothing renders below it until page reload

that's also weird. this doesn't happen for me locally ... does the PR branch deployment maybe not reflect what is currently on the branch? because the build failed or something?

> 4. Copy button doesn't have hover effect and thus doesn't look clickable. It looks a little bit weird

right, thanks. i think those styles broke with the tagless components change. should be fixed on the branch now (not sure if the deployment works).

> 5. On narrow screens (mobile) it doesn't look good

hmmmm. tough one. any idea what to do about this? maybe the buttons would need to push down to the next row on a narrow screen?

> 6. On travis-web it doesn't seem to work stable, please see the screencast

that's all wrong. when you click `customize` it doesn't show the form at all, and it also doesn't run the validation (loading indicator doesn't show up). this looks very different for me locally ...

> 7. After clicking "Trigger build" and the "Preview", I see empty configs on travis-api

again, same issue. it doesn't go to travis-yml, and therefor there's nothing to display. this is all wrong, but there's not even an exception.

> 8. Refreshing the page on "Preview" (right after (7) above) makes me stuck on that view, being unable to disable or cancel the preview mode:

also doesn't happen for me locally ...

> 9. Erasing build commit in request form makes the request fail even though the form says that commit is optional

huh, ok. for one thing, this also doesn't happen for me locally, but ... our UI kinda expects a commit sha always to be present (and so do things like the build notification email). maybe we should make it non-optional for the time being? are there other things that validate form elements in the UI so i could steal something from there?

> 10. "Trigger build" button looks like a toggle to me. It shouldn't do different actions depending on the state (e.g. enable edit mode or trigger build). Usually, when a user clicks something, clicking on the same button either does the same or "unclicks" - does the opposite action. This is what users usually expect. Running a build by this button looks unexpected. I expect there will be a lot of builds started by accident caused by this behavior.

yeah, hmmm. i tend to agree with you that it is not ideal, but then again i wouldn't 100% agree with the absoluteness of your statements :)

i'm not sure how to improve this tbh without investing a lot more space (plus time to go through the creative team, which definitely would seem necessary).

maybe a feasible compromise would be to change the original button caption to something else, like `Edit`? so the changed button caption would then indicate it's a different action? we could also, additionally, add a popup hint, maybe. (then again, the blurb says "trigger a build" etc)

> 11. The behavior after triggering a build also looks a bit unexpected to me (redirecting to requests page). I'd suggest keeping user on the same page, show a flash message and give options what they can do next, e.g. see requests or open the build. Or just stay on the same page. Automatic redirect is distracting. What happened with the changes that I just made? Are they gone? Panic!

i totally agree with this, but it's what we currently have on the trigger build popup (not something i particularly like, but yeah, it's there). i'll think about this more ...

svenfuchs commented on app/adapters/request-preview.js:
```diff
@@ -0,0 +1,25 @@
+    if (status == 400) {
+      let error = parseWithDefault(body.error_message).error || {};
+      let type = error.type;
+      let message = error.message; // TODO API should unwrap this
```
nice. done!

sorta replied:
Looks good, thanks!

svenfuchs commented on app/templates/components/request/configs-buttons.hbs:
```diff
@@ -0,0 +1,34 @@
+  {{#if this.closed}}
+    <button
+      data-test-request-configs-trigger-build="true"
+      onclick={{this.onTrigger}}
```
sounds good. i've used the `action` helper to make it consistent with the other buttons/links on the component.

sorta replied:
Makes sense! Looks good to me :)

svenfuchs commented on app/templates/components/request/configs-preview.hbs:
```diff
@@ -0,0 +1,49 @@
+    {{else}}
+      <CodeBlock
+        @language="json"
+        @code="{{this.formattedRequestConfig}}"
```
seems to work indeed. done :)

sorta replied:
Excellent, thanks! :+1:

svenfuchs commented on app/templates/components/request/configs-preview.hbs:
```diff
@@ -0,0 +1,49 @@
+
+        <CodeBlock
+          @language="json"
+          @code="{{config}}"
```
done

sorta replied:
Excellent, thanks! :+1:

svenfuchs commented on app/templates/components/request/configs-trigger-build-notice.hbs:
```diff
@@ -0,0 +1,66 @@
+    {{else if this.rejected}}
+      The build request was rejected.
+    {{else if (eq this.error 'timeout')}}
+      Processing the build request takes a very long time. TODO: now what?
```
LOL. yes. i've removed the TODO bit for now.

sorta replied:
That works! Thanks :)

svenfuchs commented on app/templates/components/request/configs-trigger-build-notice.hbs:
```diff
@@ -0,0 +1,66 @@
+        data-test-request-configs-trigger-build="true"
+        class="request-configs-button-submit button--blue"
+        title="Trigger build"
+        onclick={{this.onTrigger}}
```
done!

sorta replied:
Excellent, thanks! :+1:

svenfuchs commented on app/adapters/v3.js:
```diff
@@ -26,7 +26,7 @@ export default RESTAdapter.extend({
     for (let key in options.data) {
       let value = options.data[key];
-      if (Array.isArray(value)) {
+      if (type !== 'POST' && Array.isArray(value)) {
```
i guess we'll have to try and see :) 

without this an array with objects that is contained in a posted json body would be transformed into an array with the string `object [Object]` ...

sorta replied:
yep totally, it makes a lot of sense to me. Tests are passing and everything seems to work as far as I can tell, so I think it should be good. Thanks!

svenfuchs commented on tests/pages/trigger-build.js:
```diff
@@ -12,12 +12,14 @@ import {
-  popupTriggerLinkIsPresent: isPresent('.option-dropdown .trigger-build-anchor'),
-  openPopup: clickable('.option-dropdown .trigger-build-anchor'),
-  popupIsVisible: isVisible('.trigger-build-modal'),
+  configFormIsHidden: isHidden('[data-test-request-config-form]'),
+  configFormIsVisible: isVisible('[data-test-request-config-form]'),
```
i think i am respectfully going to leave refactoring of the tests to someone who knows what they are doing :)

sorta replied:
That's totally fair :) Sounds good! :+1:

svenfuchs commented on app/mixins/components/trigger-build.js:
```diff
@@ -0,0 +1,93 @@
+    if (data) {
+      let requestId = data.request.id;
+      let { searchDebounceRate } = config.intervals;
+      yield timeout(searchDebounceRate);
```
great catch! i think you're right. i've deleted it (and removed the import code below), and it still compiles ...

sorta replied:
Awesome! :tada:

svenfuchs commented on app/components/request/configs.js:
```diff
@@ -0,0 +1,188 @@
+import { inject as service } from '@ember/service';
+import truncate from 'travis/utils/computed';
+import CanTriggerBuild from 'travis/mixins/components/can-trigger-build';
+import TriggerBuild from 'travis/mixins/components/trigger-build';
```
done.

sorta replied:
Excellent, thanks! :+1:

svenfuchs commented on app/components/request/configs.js:
```diff
@@ -0,0 +1,188 @@
+  PREVIEW: 'preview'
+};
+
+export default Component.extend(CanTriggerBuild, TriggerBuild, {
```
done.

sorta replied:
Excellent, thanks! :+1:

svenfuchs commented on app/components/request/configs.js:
```diff
@@ -0,0 +1,188 @@
+      });
+    } else {
+      this.set('invalid', true);
+      setTimeout(() => this.set('invalid', false), 1500);
```
heh, when i've applied the change you suggested i realized we do not need that logic any more at all, as the `sha` is now optional again. so this is a left over from going back and forth between having it required or not. i've removed it.

sorta replied:
Ahh cool, that makes sense. Looks good, thanks!

svenfuchs commented on app/styles/app/modules/request/config.scss:
```diff
@@ -0,0 +1,101 @@
+    color: $dry-cement;
+    span {
+      font-family: $font-family-monospace;
+      height: 20px !important;
```
ugh. i see. it was the only way i could figure out to actually align the line numbers properly. if you'd like to give it a shot then that would be amazing!

svenfuchs commented on app/styles/app/modules/request/configs.scss:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  .request-configs-trigger-build-notice-buttons {
+    position: absolute;
+    top: 1em;
+    right: 1em;
```
i'll see what i can come up with, but yes, i think i might need help with that.

svenfuchs commented on app/adapters/request-preview.js:
```diff
@@ -0,0 +1,25 @@
+export default V3Adapter.extend({
+  buildURL: function (modelName, id, snapshot, requestType, query) {
+    return `${this._super('repo', snapshot.belongsTo('repo').id)}/request/preview`;
+  },
```
done, thanks!

svenfuchs commented on tests/acceptance/builds/config-test.js:
```diff
@@ -0,0 +1,200 @@
+    });
+
+    request.createCommit({
+      sha: 'c0ffee',
```
haha, me too! 😻

svenfuchs commented on app/components/request/configs.js:
```diff
@@ -0,0 +1,188 @@
+    if (this.closed) {
+      this.set('status', STATUSES.OPEN);
+    } else {
+      this.submitBuildRequest();
```
no, i don't think so. it's moved, yes. but this is still the top level component that handles all the actions.

svenfuchs commented on app/services/request-preview.js:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    try {
+      data.repo = yield this.store.findRecord('repo', data.repo.get('id'));
```
makes sense, done!

svenfuchs commented on app/services/request-preview.js:
```diff
@@ -0,0 +1,45 @@
+      data.repo = yield this.store.findRecord('repo', data.repo.get('id'));
+      const record = this.store.createRecord('request-preview', data);
+      yield record.save();
+      this.setProperties({ record: record, loaded: true });
```
hmmmm, i believe we've tried this before, or do i misremember?

i think the issue is that the `request/configs` component needs to be able to `reset` this property (i.e. set it to `false`) when the user clicks `cancel` and closes the component. the property previously was defined on the component, rather than the service in order to not keep this state here ... but then after some discussion with Sophie (if i remember correctly?) we've decided to move it here.

svenfuchs replied:
fwiw there _is_ a way to reset the state on the service, and this is now resolved here: https://github.com/travis-ci/travis-web/commit/f6d98997905ecbabb29a1fc575bde48166d7aa5a

svenfuchs commented on app/services/trigger-build.js:
```diff
@@ -0,0 +1,54 @@
+
+  submit: task(function* (data) {
+    try {
+      data.commit = this.store.createRecord('commit', data.commit);
```
done!

artursmirnov commented:
Hey @svenfuchs 👋 It seems like this PR is going to be released soon 🎉 Although it haven't been updated from `master` for quite a while and it has conflicts. Would you like to prepare it on your own or would you like me to take care of the preparation?
